src/random.cpp

```diff
@@ -33,41 +33,21 @@ POSSIBILITY OF SUCH DAMAGE.
 #include "libtorrent/config.hpp"
 #include "libtorrent/random.hpp"
 
-#include <random>
-
 namespace libtorrent
 {
-	using std::random_device;
-	using std::mt19937;
-	using std::uniform_int_distribution;
-
-#ifdef TORRENT_BUILD_SIMULATOR
-
-	std::uint32_t random(std::uint32_t max)
+	namespace aux
 	{
+#ifdef TORRENT_BUILD_SIMULATOR
 		// make sure random numbers are deterministic. Seed with a fixed number
-		static mt19937 random_engine(0x82daf973);
-		return uniform_int_distribution<std::uint32_t>(0, max)(random_engine);
-	}
-
+		std::mt19937 random_engine(0x82daf973);
 #else
-
-	std::uint32_t random(std::uint32_t max)
-	{
-		// TODO: versions prior to msvc-14 (visual studio 2015) do
-		// not generate thread safe initialization of statics
-		static random_device dev;
-		static mt19937 random_engine(dev());
-		return uniform_int_distribution<std::uint32_t>(0, max)(random_engine);
+		static std::random_device dev;
+		std::mt19937 random_engine(dev());
```

Comment:
this is problematic, because the initialization order of global variables aren't defined.
It goes beyond this too though, if any global initializer happens to call aux::random_shuffle, it would end up potentially using "random_engine" before it was initialized.

The simplest way to solve this problem is to have a function that returns the random number generator instead, and it can have a local static variable which it returns a reference to. That way, whoever calls that function first will be responsible to initialize the object, and it would be guaranteed to be valid by the time it's first used. random() have this effect for instance.